src/Microsoft.TemplateEngine.Cli/TemplateUpdateChecker.cs

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.TemplateEngine.Abstractions;
+using Microsoft.TemplateEngine.Abstractions.TemplateUpdates;
+
+namespace Microsoft.TemplateEngine.Cli
+{
+    public class TemplateUpdateChecker
+    {
+        public TemplateUpdateChecker(IEngineEnvironmentSettings environmentSettings)
+        {
+            _environmentSettings = environmentSettings;
+        }
+
+        private readonly IEngineEnvironmentSettings _environmentSettings;
+        // Maps the install unit descriptor factory ids to the corresponding updaters.
+        private Dictionary<Guid, IUpdater> _factoryIdToUpdaterMap;
+
+        public async Task<IReadOnlyList<IUpdateUnitDescriptor>> CheckForUpdatesAsync(IReadOnlyList<IInstallUnitDescriptor> installUnitsToCheck)
+        {
+            EnsureFactoryToUpdaterMapping();
+            Dictionary<Guid, List<IInstallUnitDescriptor>> installUnitsToCheckForUpdates = new Dictionary<Guid, List<IInstallUnitDescriptor>>();
+            List<IUpdateUnitDescriptor> updateDescriptors = new List<IUpdateUnitDescriptor>();
+
+            // collect the descriptors by their factoryId, ignoring descriptors that don't have corresponding factories or updaters.
+            foreach (IInstallUnitDescriptor descriptor in installUnitsToCheck)
+            {
+                if (_factoryIdToUpdaterMap.TryGetValue(descriptor.FactoryId, out IUpdater updater))
+                {
+                    if (!installUnitsToCheckForUpdates.TryGetValue(descriptor.FactoryId, out List<IInstallUnitDescriptor> updateList))
+                    {
+                        updateList = new List<IInstallUnitDescriptor>();
+                        installUnitsToCheckForUpdates[descriptor.FactoryId] = updateList;
+                    }
+
+                    updateList.Add(descriptor);
+                }
+            }
+
+            // check for updates
+            foreach (KeyValuePair<Guid, List<IInstallUnitDescriptor>> descriptorsForType in installUnitsToCheckForUpdates)
+            {
+                if (_factoryIdToUpdaterMap.TryGetValue(descriptorsForType.Key, out IUpdater updater))
```

Comment:
If there's no updater for a descriptor, is it worth to log something? Or is it super easy to figure out if an update didn't happen as there was no updater for it?

Reply:
We should probably report something to the user, so they know certain template packs didn't get checked for updates. Anything with a descriptor should probably have an updater, but it's not required.

Reply:
I added the checking above this loop. The first loop tries to map the descriptors to updaters, and now notes the ones without updaters. They get reported all at once in the new `if (descriptorsWithoutUpdaters.Count > 0)`. The second loop is using the mapping from the first loop, so anything in installUnitsToCheckForUpdates is guaranteed to have an updater.